### PR TITLE
Initial support for Link Headers

### PR DESCRIPTION
**Pull request by grandelc**
- Added the PageInfo model
- Parsing the Link Headers from the response to retrieve the page info
- Removed the unused iterator functio from the CollectionEntity
- Added the page info retrieval into the CollectionEntity class

### CODE REVIEW

File: src/Services/Client.php [hunk 1]
```diff
@@ -238,6 +242,12 @@ private function sendRequestToShopify(Request $request, array $cookies = [], $pa
             $this->requestHookInterface->afterRequest($response);
         }
 
+        if ($response && $linkHeader = $response->getHeader('Link')) {
```
Comment (JoshuaKlassen) [inline src/Services/Client.php]: Do you mean to be setting $linkHeader here, or is it meant to be a comparison?

File: src/Services/Client.php [hunk 2]
```diff
@@ -305,6 +315,12 @@ private function getRedirectResponseFromShopify(Request $request)
...
```
Comment (JoshuaKlassen) [inline src/Services/Client.php]: Same here, is this intentional? It's a bit confusing reading it

**Comment by wfotiuk** [comment]
What version of PHP are we making this for? Can we add strict types to function params here?

**Comment by wfotiuk** [comment]
Also I don't see any tests here, is it possible to add some?

**Comment by grandelc** [comment]
@wfotiuk Julian and I are working on getting some tests setup 👍

A PHP version hasn't been specified. Open including strict types but perhaps as a separate MR?

File: src/Services/Client.php [hunk 3]
```diff
@@ -330,4 +346,67 @@ private function validateRedirectLocation($location)
...
+    {
+        $headerLinks = $this->parseHeader($linkHeader);
+        $nextPageInfo = '';
+        $previousPageInfo = '';
+        if (is_array($headerLinks) && isset($headerLinks['next'])) {
+            $nextUrlComponents = parse_url($headerLinks['next']);
+            parse_str($nextUrlComponents['query'], $nextParams);
+            $nextPageInfo = $nextParams['page_info'] ? $nextParams['page_info'] : '';
+        }
+        if (is_array($headerLinks) && isset($headerLinks['prev'])) {
```
Comment (JoshuaKlassen) [inline src/Services/Client.php]: `$headerLinks` should always be an array right? I think then you could remove it, or maybe nest the 'next' and 'prev' checks inside? I'm just thinking for readability - what do you think?

**Comment by CoryMik** [comment]
@grandelc @wfotiuk we have a couple applications using this that are still on php 5.6. I'm not sure if we want to restrict the version of php specifically for strict typing.

**Comment by wfotiuk** [comment]
I thought we were all off of php 5.6 at BOLD. No biggie though I do think maybe we should do a major and add strict typing at one point though, LGTM!

File: src/Services/Client.php (hunk 1 above)
Comment (grandelc) [reply src/Services/Client.php]: I agree its confusing, we'll fix it!

File: src/Services/Client.php (hunk 2 above)
Comment (grandelc) [reply src/Services/Client.php]: I agree its confusing, we'll fix it!

File: src/Services/Client.php (hunk 3 above)
Comment (grandelc) [reply src/Services/Client.php]: Thats fair it should always be an array, we will remove it!

File: src/Services/Client.php [hunk 4]
```diff
@@ -37,6 +39,8 @@ class Client
     /** @var RateLimitKeyGenerator */
     protected $rateLimitKeyGenerator;
 
+    protected $pageInfo;
```
Comment (Wryte) [inline src/Services/Client.php:42]: This var should have a doc block to type hint.

File: src/Services/Client.php [hunk 5]
```diff
@@ -276,6 +287,14 @@ private function getRedirectResponseFromShopify(Request $request)
                     $result = $this->validateRedirectLocation($headers[$locationIndex][0]);
                 }
             }
+
+            $linkHeader = $response->getHeader('Link');
+
```
Comment (Wryte) [inline src/Services/Client.php:292]: Small thing, extra whitespace compared to other added if block above.

File: src/Services/Client.php [hunk 6]
```diff
@@ -330,4 +349,67 @@ private function validateRedirectLocation($location)
 
         return $redirectLocation;
     }
+
+    public function getPageInfo()
```
Comment (Wryte) [inline src/Services/Client.php:353]: Add doc block to this method and parseHeader as well.

File: src/Services/Client.php [hunk 7]
```diff
@@ -330,4 +349,67 @@ private function validateRedirectLocation($location)
...
+     * Returns an array of the next and previous page info parameters extracted from an array of next and previous links.
+     *
+     * @param $linkHeader
+     */
+    private function parsePageInfo($linkHeader)
+    {
+        $headerLinks = $this->parseHeader($linkHeader);
+        $nextPageInfo = '';
+        $previousPageInfo = '';
+        if (isset($headerLinks['next'])) {
```
Comment (Wryte) [inline src/Services/Client.php:389]: There are no test cases for when $headerLinks['next'] or $headerLinks['prev'] are not set.

**Comment by grandelc** [comment]
What are peoples thoughts on the name for the Collection service class? should we rename it? Keep it as is? 

I followed the naming convention of the service classes being based off the Shopify API names. However Collection in particular can easily conflict with the Illuminate Collection class. 

@Wryte @CoryMik @wfotiuk @JoshuaKlassen

File: src/Services/Client.php (hunk 4 above)
Comment (BoldJulian) [reply src/Services/Client.php:43]: Updated in latest commit

File: src/Services/Client.php (hunk 6 above)
Comment (BoldJulian) [reply src/Services/Client.php:356]: Updated in latest commit

File: src/Services/Client.php (hunk 7 above)
Comment (BoldJulian) [reply src/Services/Client.php:399]: I added some more tests in the latest commit, will need to be reviewed.

File: src/Services/Client.php (hunk 5 above)
Comment (BoldJulian) [reply src/Services/Client.php]: Updated in latest commit, removed the extra whitespace.

**Comment by JoshuaKlassen** [comment]
Hmm - its good to keep a consistent codebase so keeping the naming convention is good. But I can see the confusion

Can we just alias it to `CollectionService` in whatever files use it?
```
Use services\Collection as CollectionService;
```
That might help prevent the confusion

File: src/Services/CollectionEntity.php [hunk 8]
```diff
@@ -5,33 +5,31 @@
...
-                    break;
-                }
-            }
-        }
+        return $this->client->getPageInfo();
     }
 
     /**
-     * @param $params
+     * @param $pageInfo
```
Comment (Wryte) [inline src/Services/CollectionEntity.php:23]: There's just a few more doc block fixes that I think will avoid confusion for users of this package.

This pageInfo param is a string right? Add the hint.

In this same class the getPageInfo method returns an actual PageInfo instance right? You can change the return type to PageInfo then?

So when someone goes to use this the would do

```php
$pageInfo = $collectionEntityService->getPageInfo();
$nextPageCollections = $collectionEntityService->getByPageInfo($pageInfo->getNext());
```

I think having string as a the type on this method will avoid the confusion that you are able to pass in a PageInfo object.

File: tests/CollectionTest.php [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+            "handle": "whole-store",
+            "title": "Whole Store",
+            "updated_at": "2020-03-28T21:20:10-04:00",
+            "published_at": "2020-03-28T21:18:40-04:00",
+            "sort_order": "best-selling",
+            "template_suffix": "",
+            "products_count": 269,
+            "collection_type": "custom",
+            "published_scope": "web",
+            "body_html": ""
```
Comment (JoshuaKlassen) [inline tests/CollectionTest.php:46]: Suggestion: this could be moved out into a dataprovider, and the `createShopifyCollectionEntity` function could use the data from the provider - that we we can add multiple test cases easily and don't have to duplicate the data for 2 functions.

Comment (Wryte) [reply tests/CollectionTest.php:46]: And if we're talking dataproviders (which, why not?) the tests added in ClientTest are begging for one also. 👍

File: tests/Services/ClientTest.php [hunk 10]
```diff
@@ -182,6 +183,204 @@ public function testClientWillApplyPostHeadersProperly()
...
+        $pageInfo = $this->client->getPageInfo();
+
+        $parsedNextPageInfo = $pageInfo->getNext();
+        $parsedPrevPageInfo = $pageInfo->getPrev();
+
+        $this->assertEquals($expectedNextPageInfo, $parsedNextPageInfo);
+        $this->assertEquals($expectedPrevPageInfo, $parsedPrevPageInfo);
+    }
+
+    public function testClientParsePageInfoNoNext()
```
Comment (JoshuaKlassen) [inline tests/Services/ClientTest.php:250]: suggestion: These 3 tests could be made into a single test with a dataprovider that has all 3 cases (both prev and next, just prev, just next). You can provide the expected result in the data provider as well.

File: tests/Services/ClientTest.php [hunk 11]
```diff
@@ -182,6 +183,204 @@ public function testClientWillApplyPostHeadersProperly()
...
+            $this->mockShopBaseInfo, $this->mockShopAccessInfo, $mockHttpClient,
+            $this->mockRequestHookInterface
+        );
+
+        $parsedLinks = $this->callMethod($this->client, 'parseHeader', array($linkHeader));
+
+        $this->assertEquals($expectedHeaderLinks, $parsedLinks);
+    }
+
+    public function testClientParseHeaderNoNext()
```
Comment (JoshuaKlassen) [inline tests/Services/ClientTest.php:338]: suggestion: Same as above, these 3 tests could be made into a single test with a dataprovider that has all 3 cases (both prev and next, just prev, just next). You can provide the expected result in the data provider as well.

**Comment by grandelc** [comment]
Will be separating the support for the Collection service as a separate pull request

File: src/Services/CollectionEntity.php (hunk 8 above)
Comment (BoldJulian) [reply src/Services/CollectionEntity.php]: I agree, this does clear up some confusion! 
I've made the changes on my latest commit, should be good now.

File: tests/Services/ClientTest.php (hunk 10 above)
Comment (BoldJulian) [reply tests/Services/ClientTest.php]: Good suggestion, updated to use a dataprovider for all 3 test cases in my latest commit.

File: tests/Services/ClientTest.php (hunk 11 above)
Comment (BoldJulian) [reply tests/Services/ClientTest.php]: Good suggestion again :), updated to use a dataprovider for all 3 test cases in my latest commit. Let me know if the latest version of the tests look good!

Comment (JoshuaKlassen) [reply tests/Services/ClientTest.php]: Why can't I figure out how to close these comments?
